compiler/rustc_codegen_llvm/src/llvm_util.rs

```diff
@@ -221,6 +221,26 @@ pub fn target_cpu(sess: &Session) -> &str {
     handle_native(name)
 }
 
+pub fn handle_native_features(sess: &Session) -> Vec<String> {
+    match sess.opts.cg.target_cpu {
+        Some(ref s) => {
+            if s != "native" {
+                return vec![];
+            }
+
+            let ptr = unsafe { llvm::LLVMGetHostCPUFeatures() };
+            let str = unsafe { CStr::from_ptr(ptr).to_string_lossy() };
```

Comment:
style nit: override of the `str` seems somewhat iffy.

I also think this should just `bug!` out if the string doesn't convert to utf8 rather than replacing characters with replacement characters.

This should also check for `ptr` being null before dereferencing it.

Finally, another style nit, although I could go either way on it: all of these unsafe expressions should share the same unsafe block. As thus I would write this sippet as follows:

```rust
let features_string = unsafe {
    let ptr = llvm::LLVMGetHostCPUFeatures();
    let features_string = if !ptr.is_null() {
        String::from(CStr::from_ptr(ptr).to_str().unwrap_or_else(|| {
            bug!("LLVM returned a non-utf8 features string");
        }))
    } else {
        bug!("could not allocate host CPU features, LLVM returned a `null` string")
    };
    LLVMDisposeMessage(ptr);
    features_string
};
// split here...
```

One thing this achieves is making it more clear that there are non-trivial invariants shared between the various unsafe calls in this function.